### PR TITLE
[doc][api] api policy lint check for data team

### PR DESCRIPTION
**Pull request by can-anyscale**
Add a lint to check for the first api policy: every public API needs to be documented. This PR adds the check for `data` team.

There are a list of APIs that are not following the rules yes, I'll remove them from the `whitelist` as we are fixing them.

Test:
- CI

### CODE REVIEW

File: ci/ray_ci/doc/api.py [hunk 1]
```diff
@@ -117,15 +118,17 @@ def get_canonical_name(self) -> str:
...
+        return self.name
 
     def _is_private_name(self) -> bool:
         """
         Check if this API has a private name. Private names are those that start with
         underscores.
         """
         name_has_underscore = self.name.split(".")[-1].startswith("_")
-        is_internal = ".internal." in self.name
+        is_internal = "._internal." in self.name
```
Comment (can-anyscale) [inline ci/ray_ci/doc/api.py:131]: the internal package is named as `_internal`, fix this here

File: ci/ray_ci/doc/api.py [hunk 2]
```diff
@@ -117,15 +118,17 @@ def get_canonical_name(self) -> str:
                 return self.name
             attribute = getattr(attribute, token)
 
-        return f"{attribute.__module__}.{attribute.__qualname__}"
+        if inspect.isclass(attribute) or inspect.isfunction(attribute):
```
Comment (can-anyscale) [inline ci/ray_ci/doc/api.py:121]: also fix this, only class or function/method APIs can have canonical name

File: .buildkite/lint.rayci.yml [hunk 3]
```diff
@@ -20,17 +20,25 @@ steps:
...
+    key: lint-api-annotations
     instance_type: medium
     depends_on:
       - oss-ci-base_build
     commands:
-      - ./ci/lint/lint.sh {{matrix}}
+      - ./ci/lint/lint.sh api_annotations
+    job_env: oss-ci-base_build
+
+  - label: ":lint-roller: lint: {{matrix}} api policy"
```
Comment (aslonnie) [inline .buildkite/lint.rayci.yml:32]: `api_policy_check`? or change the other one to `api annotations` ?

**Review by aslonnie — APPROVED** [review]

**Comment by can-anyscale** [comment]
use `api_policy_check`

**Comment by can-anyscale** [comment]
CC: @bveeramani for FYI that I'm merging this; this will block creation of additional APIs that are not documented; let me know if the team is get blocked by this in anyway, thankks ;)

File: ci/ray_ci/doc/cmd_check_api_discrepancy.py [hunk 4]
```diff
@@ -1,23 +1,88 @@
 import click
+from typing import Dict, Set
 
 from ci.ray_ci.doc.module import Module
+from ci.ray_ci.doc.autodoc import Autodoc
+from ci.ray_ci.doc.api import API
+
+TEAM_API_CONFIGS = {
+    "data": {
+        "head_modules": {"ray.data", "ray.data.grouped_data"},
```
Comment (bveeramani) [inline ci/ray_ci/doc/cmd_check_api_discrepancy.py:10]: Why do we include `ray.data.grouped_data` in particular and not another module like `ray.data.read_api`?

**Pull request merged**